Refactor generate_ci_workflows

zhouzhuojie opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #61886
* __->__ #61879

Refactor generate_ci_workflows to support CI dispatcher. This is the first step to refactor the workflow into a dataclass with some validation and OOP.

Verified that the output is the same:

```
.github/scripts/generate_ci_workflows.py
git status
```

Differential Revision: [D29808365](https://our.internmc.facebook.com/intern/diff/D29808365)

zhouzhuojie committed:
Refactor generate_ci_workflows

[ghstack-poisoned]

zhouzhuojie commented on requirements.txt:
```diff
@@ -7,5 +7,6 @@ pyyaml
 requests
 setuptools
 six
+types-dataclasses
```
required at dev time for mypy to support dataclasses

zhouzhuojie commented on .github/scripts/generate_ci_workflows.py:
```diff
@@ -1,123 +1,98 @@
-    }
-
+@dataclass
+class CIWorkflow:
```
consolidating into a single dataclass to support some validation and oop

samestep replied:
hm I'm not sure I really like this actually... it seems like it just turns the code into a more complex form that involves more `if` statements; I'm guessing this refactoring is necessary for #61886 in some way?

samestep replied:
although I do like how this reduces the risk of typos that we previously had from duplicating variable names and their corresponding field names in the `PyTorchWindowsWorkflow` and `PyTorchLinuxWorkflow` functions

zhouzhuojie replied:
originally, I was doing the split work that but just too many duplicate code between PyTorchWindowsWorkflow and PyTorchLinuxWorkflow, and no types for PyTorchWorkflow, which makes me think that it's worth to just DRY it up and have validation for both.

and yes, it helps to reduce the risk of typos and we don't need to remember to add parameters for both, imagine adding macos workflows and we got 3 at that point.

zhouzhuojie committed:
Update on "Refactor generate_ci_workflows"




[ghstack-poisoned]

samestep commented on .github/scripts/generate_ci_workflows.py:
```diff
@@ -1,123 +1,107 @@
-
 WINDOWS_CPU_TEST_RUNNER = "windows.4xlarge"
 WINDOWS_CUDA_TEST_RUNNER = "windows.8xlarge.nvidia.gpu"
+WINDOWS_RUNNERS = [
+    WINDOWS_CPU_TEST_RUNNER,
+    WINDOWS_CUDA_TEST_RUNNER,
+]
 
+LINUX_CPU_TEST_RUNNER = "linux.2xlarge"
+LINUX_CUDA_TEST_RUNNER = "linux.8xlarge.nvidia.gpu"
+LINUX_RUNNERS = [
+    LINUX_CPU_TEST_RUNNER,
+    LINUX_CUDA_TEST_RUNNER,
+]
```
could you make these sets instead of lists? just to clarify that they're used for membership checks and their order doesn't matter

samestep commented on .github/scripts/generate_ci_workflows.py:
```diff
@@ -1,123 +1,107 @@
-        "is_libtorch": is_libtorch,
-        "exclude_test": is_libtorch or exclude_test,   # libtorch is build only
-    }
+    def assert_valid(self) -> None:
+        assert self.arch in ['linux', 'windows'], f"invalid arch: {self.arch}, must be one of ['linux','windows']"
```
rather than doing this with a dynamic check, why not just make the type of `arch` be `Literal["linux", "windows"]` instead of just `str`?

zhouzhuojie committed:
Update on "Refactor generate_ci_workflows"


Refactor generate_ci_workflows to support CI dispatcher. This is the first step to refactor the workflow into a dataclass with some validation and OOP.

Verified that the output is the same:

```
.github/scripts/generate_ci_workflows.py
git status
```


[ghstack-poisoned]

zhouzhuojie commented:
@zhouzhuojie has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29808365).

samestep commented on .github/scripts/generate_ci_workflows.py:
```diff
@@ -1,123 +1,106 @@
-
 WINDOWS_CPU_TEST_RUNNER = "windows.4xlarge"
 WINDOWS_CUDA_TEST_RUNNER = "windows.8xlarge.nvidia.gpu"
+WINDOWS_RUNNERS = set([
+    WINDOWS_CPU_TEST_RUNNER,
+    WINDOWS_CUDA_TEST_RUNNER,
+])
```
if you want, you can just use a set literal (doesn't matter though)
```suggestion
WINDOWS_RUNNERS = {
    WINDOWS_CPU_TEST_RUNNER,
    WINDOWS_CUDA_TEST_RUNNER,
}
```

samestep commented on .github/scripts/generate_ci_workflows.py:
```diff
@@ -1,123 +1,106 @@
 
+LINUX_CPU_TEST_RUNNER = "linux.2xlarge"
+LINUX_CUDA_TEST_RUNNER = "linux.8xlarge.nvidia.gpu"
+LINUX_RUNNERS = set([
+    LINUX_CPU_TEST_RUNNER,
+    LINUX_CUDA_TEST_RUNNER,
+])
```
same here
```suggestion
LINUX_RUNNERS = {
    LINUX_CPU_TEST_RUNNER,
    LINUX_CUDA_TEST_RUNNER,
}
```

zhouzhuojie committed:
Update on "Refactor generate_ci_workflows"


Refactor generate_ci_workflows to support CI dispatcher. This is the first step to refactor the workflow into a dataclass with some validation and OOP.

Verified that the output is the same:

```
.github/scripts/generate_ci_workflows.py
git status
```

Differential Revision: [D29808365](https://our.internmc.facebook.com/intern/diff/D29808365)

[ghstack-poisoned]

zhouzhuojie commented:
@zhouzhuojie has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29808365).

seemethere commented on .github/scripts/generate_ci_workflows.py:
```diff
@@ -1,123 +1,106 @@
 WINDOWS_WORKFLOWS = [
-    PyTorchWindowsWorkflow(
+    CIWorkflow(
+        arch="windows",
```
Should we just make this value a constant? To avoid people mistyping this?

samestep replied:
this is solved by the `Arch = Literal["windows", "linux"]` definition; see https://github.com/pytorch/pytorch/pull/61879#discussion_r673427592

seemethere reviewed (approved):
one minor nit otherwise lgtm

zhouzhuojie committed:
Update on "Refactor generate_ci_workflows"


Refactor generate_ci_workflows to support CI dispatcher. This is the first step to refactor the workflow into a dataclass with some validation and OOP.

Verified that the output is the same:

```
.github/scripts/generate_ci_workflows.py
git status
```

Differential Revision: [D29808365](https://our.internmc.facebook.com/intern/diff/D29808365)

[ghstack-poisoned]

zhouzhuojie commented:
@zhouzhuojie has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29808365).
